File: code/python/src/runbench.py
import argparse
from dataclasses import dataclass
import subprocess

import runsynth


@dataclass
class Opts():
    time_limit: int
    mem_limit: int
    output_dir: str
    files: [str]
    num_examples: int
    num_consts_int: int
    num_consts_str: int


def runbench(opts: Opts):
    directory = opts.output_dir
    if not os.path.exists(directory):
        os.makedirs(directory)

    for file in opts.files:
        subprocess.run(
            'runsolver',
            [
                '--wall-clock-limit',
                opts.time_limit,
                '--mem-limit',
                opts.mem_limit * 1000,  # megabytes
                '--solver-data',
                f'{opts.outputDir}/{file}.out',
                '--timestamp',
                'outsynth2',
                '--examples',
                opts.num_examples,
                '--constants',
                opts.num_consts,
                file
            ])


def add_res_args(parser):
    parser.add_argument(
        '-t',
        '--time-limit',
        metavar='TIME-LIMIT',
        default=1,
        type=int,
        help='Time limit (wall-clock time) in seconds')
    parser.add_argument(
        '-m',
        '--mem-limit',
        default=12,
        type=int,
        help='Memory limit in gigabytes')


def add_dir_arg(parser):
    parser.add_argument(
        '-d',
        '--output-dir',
        metavar='OUTPUT-DIR',
        help='Directory where to save the benchmarks')


def add_files_arg(parser):
    parser.add_argument(
        'file',
        metavar='FILE',
        type=str,
        nargs='+',
        help='The file with the examples')


def argparser():
    parser = argparse.ArgumentParser(
        description='outsynth2 benchmark runner')
    add_res_args(parser)
    add_dir_arg(parser)
    add_files_arg(parser)
    runsynth.add_example_arg(parser)
    runsynth.add_consts_args(parser)
    return parser


if __name__ == '__main__':
    parser = argparser()
    args = parser.parse_args()
    opts = Opts(
        time_limit=args.time_limit,
        mem_limit=args.mem_limit,
        output_dir=args.output_dir,
        file=args.file,
        num_examples=args.examples,
        num_consts_int=args.consts_int,
        num_consts_str=args.consts_str)
    runsynth(opts)

File: code/whole/src/RunBench.hs
module RunBench where

import           Control.Monad
import           System.Directory
import           System.Process

import           Options.Applicative

import qualified RunSynth            as S hiding (main)


data Opts = Opts
  Int -- ^ wall-clock time limit in seconds
  Int -- ^ memory limit in gigabytes
  FilePath -- ^ output directory
  [S.OptFile] -- ^ files with examples
  S.OptNumExamples -- ^ number of examples to take from each benchmark
  S.OptNumConsts -- ^ number of constants to use


timeLimit (Opts t _ _ _ _ _) = t
memLimit (Opts _ m _ _ _ _) = m
outputDir (Opts _ _ d _ _ _) = d
files (Opts _ _ _ fs _ _) = fs
numExamples (Opts _ _ _ _ (S.OptNumExamples e) _) = e
numConsts (Opts _ _ _ _ _ (S.OptNumConsts c)) = c


opts :: Parser Opts
opts = Opts <$>
  optTimeLimit <*>
  optMemLimit <*>
  optOutputDir <*>
  some S.optFile <*>
  S.optNumExamples <*>
  S.optNumConsts


optTimeLimit :: Parser Int
optTimeLimit = option auto (long "time-limit" <>
                            short 't' <>
                            metavar "TIME-LIMIT" <>
                            showDefault <>
                            value 1 <>
                            help "Time limit (wall-clock time) in seconds")


optMemLimit :: Parser Int
optMemLimit = option auto (long "mem-limit" <>
                           short 'm' <>
                           metavar "MEM-LIMIT" <>
                           showDefault <>
                           value 12 <>
                           help "Memory limit in gigabytes")


optOutputDir :: Parser FilePath
optOutputDir = strOption (long "output-dir" <>
                          short 'd' <>
                          metavar "OUTPUT-DIR" <>
                          help "Directory where to save the benchmarks")


optsInfo :: ParserInfo Opts
optsInfo = info (opts <**> helper)
           (fullDesc <> header "outsynth2 benchmark runner")


ensureRunsolver :: IO ()
ensureRunsolver = void $ readProcess "which" ["runsolver"] ""


runBench :: Opts -> IO ()
runBench opts = do
  ensureRunsolver

  createDirectoryIfMissing True (outputDir opts)

  forM_ (files opts) $ \(S.OptFile file) ->
    readProcess "runsolver" [ "--wall-clock-limit"
                            , show $ timeLimit opts
                            , "--mem-limit"
                            , show $ memLimit opts * 1000 -- megabytes
                            , "--solver-data"
                            , outputDir opts <> "/" <> file <> ".out"
                            , "--timestamp"
                            , "outsynth2"
                            , "--examples"
                            , show $ numExamples opts
                            , "--constants"
                            , show $ numConsts opts
                            , file
                            ] ""


main :: IO ()
main = runBench =<< execParser optsInfo
